Miniscript integration

darosior opened the pull request:
Miniscript is a language for writing (a subset of) Bitcoin Scripts in a structured way.

Miniscript permits:
- To safely extend the Output Descriptor language to many more scripting features thanks to the typing system (composition).
- Static analysis of spending conditions, maximum spending cost of each branch, security properties, third-party malleability.
- General satisfaction of any correctly typed ("valid" [0]) Miniscript. The satisfaction itself is also analyzable.
- To extend the possibilities of external signers, because of all of the above and since it carries enough metadata.

Miniscript guarantees:
- That for any statically-analyzed as "safe" [0] Script, a witness can be constructed in the bounds of the consensus and standardness rules (standardness complete).
- That unless the conditions of the Miniscript are met, no witness can be created for the Script (consensus sound).
- Third-party malleability protection for the satisfaction of a sane Miniscript, which is too complex to summarize here.

For more details around Miniscript (including the specifications), please refer to the [website](https://bitcoin.sipa.be/miniscript/).

Miniscript was designed by Pieter Wuille, Andrew Poelstra and Sanket Kanjalkar.
This PR is an updated and rebased version of #16800. See [the commit history of the Miniscript repository](https://github.com/sipa/miniscript/commits/master) for details about the changes made since September 2019 (TL;DR: bugfixes, introduction of timelock conflicts in the type system, `pk()` and `pkh()` aliases, `thresh_m` renamed to `multi`, all recursive algorithms were made non-recursive).

This PR is also the first in a series of 3:
- The first one (here) integrates the backbone of Miniscript.
- The second one (#24148) introduces support for Miniscript in Output Descriptors, allowing for watch-only support of Miniscript Descriptors in the wallet.
- The third one (#24149) implements signing for these Miniscript Descriptors, using Miniscript's satisfaction algorithm. 

Note to reviewers:
- Miniscript is currently defined only for P2WSH. No Taproot yet.
- Miniscript is different from the policy language (a high-level logical representation of a spending policy). A policy->Miniscript compiler is not included here.
- The fuzz target included here is more interestingly extended in the 3rd PR to check a script's satisfaction against `VerifyScript`. I think it could be further improved by having custom mutators as we now have for multisig (see https://github.com/bitcoin/bitcoin/issues/23105). A minified corpus of Miniscript Scripts is available at https://github.com/bitcoin-core/qa-assets/pull/85.

[0] We call "valid" any correctly-typed Miniscript. And "safe" any sane Miniscript, ie one whose satisfaction isn't malleable, which requires a key for any spending path, etc..

sipa committed:
Miniscript: type system, script creation, text notation, tests

More information about Miniscript can be found at https://bitcoin.sipa.be/miniscript/ (the
website source is hosted at https://github.com/sipa/miniscript/).
This commit defines all fragments, their composition, parsing from
string representation and conversion to Script.

Co-Authored-By: Antoine Poinsot <darosior@protonmail.com>
Co-Authored-By: Sanket Kanjalkar <sanket1729@gmail.com>
Co-Authored-By: Samuel Dobson <dobsonsa68@gmail.com>

sipa committed:
Miniscript: ops limit and stack size computation

Co-Authored-By: Antoine Poinsot <darosior@protonmail.com>

sipa committed:
Miniscript: conversion from script

Co-Authored-By: Antoine Poinsot <darosior@protonmail.com>
Co-Authored-By: Samuel Dobson <dobsonsa68@gmail.com>

darosior committed:
script: make IsPushdataOp non-static

We'll need it for Miniscript

darosior committed:
script: move CheckMinimalPush from interpreter to script.h

It is used by Miniscript.

darosior committed:
fuzz: add a fuzz target for Miniscript decoding from Script

MarcoFalke commented on src/script/miniscript.h:
```diff
@@ -0,0 +1,1563 @@
+        // and the string representations of its child nodes, the string representation of the node.
+        auto upfn = [&ctx](bool wrapped, const Node& node, Span<std::string> subs) -> std::optional<std::string> {
+            std::string ret = wrapped ? ":" : "";
+            std::stringstream k_sstr;
```
Would it compile if you remove this and use `::ToString(node.k)` instead?

darosior replied:
:man_facepalming: Yes. Done.

darosior commented:
Arg, the unit tests (`miniscript_tests`) get OOM killed in the CI (https://github.com/bitcoin/bitcoin/pull/24147/checks?check_run_id=4936250775, https://cirrus-ci.com/task/6650942870257664).

Sjors commented:
Concept ACK. I wonder if it makes sense to start out with a small subset of miniscript, so reviewers can focus on the implementation rather than on completeness.

meshcollider commented:
Strong concept ACK

sipa commented on src/test/miniscript_tests.cpp:
```diff
@@ -0,0 +1,407 @@
+    g_testdata.reset();
+}
+
+BOOST_AUTO_TEST_CASE(random_tests)
```
This test, along with all the RandomNode logic, is probably much more useful if converted to a fuzz test (which doesn't need to construct a realistic/useful probability model, but can just choose based on inputs). Let me know if you want any help with that.

sipa commented on src/script/miniscript.h:
```diff
@@ -549,10 +551,10 @@ struct Node {
 
     //! Check whether this node is valid at all.
-    bool IsValid() const { return !(GetType() == ""_mst); }
+    bool IsValid() const { return !(GetType() == ""_mst) && ScriptSize() <= MAX_STANDARD_P2WSH_SCRIPT_SIZE; }
```
In commit "Miniscript: conversion from script"

I think this change (as well as the one to `IsValidTopLevel` below) is in the wrong commit.

MarcoFalke commented on src/script/script.h:
```diff
@@ -429,6 +429,20 @@ class CScript : public CScriptBase
+        return *this;
+    }
+
+    friend CScript operator+(const CScript& a, const CScript& b)
```
I don't really like this. Does this also enable `CScript{} + CScriptBase{}`? If yes, that sounds bad, because you likely don't want to append data without pushing it.

Even if it doesn't, is there any downside in having a named function for this? The existing `Cat` might even work for this?

darosior replied:
Indeed, just using `Cat` does work.
```diff
diff --git a/src/script/miniscript.h b/src/script/miniscript.h
index c67a3c4ea0..88fa4730cd 100644
--- a/src/script/miniscript.h
+++ b/src/script/miniscript.h
@@ -446,7 +446,7 @@ public:
                 case NodeType::RIPEMD160: return CScript() << OP_SIZE << 32 << OP_EQUALVERIFY << OP_RIPEMD160 << node.data << (verify ? OP_EQUALVERIFY : OP_EQUAL);
                 case NodeType::HASH256: return CScript() << OP_SIZE << 32 << OP_EQUALVERIFY << OP_HASH256 << node.data << (verify ? OP_EQUALVERIFY : OP_EQUAL);
                 case NodeType::HASH160: return CScript() << OP_SIZE << 32 << OP_EQUALVERIFY << OP_HASH160 << node.data << (verify ? OP_EQUALVERIFY : OP_EQUAL);
-                case NodeType::WRAP_A: return (CScript() << OP_TOALTSTACK) + std::move(subs[0]) + (CScript() << OP_FROMALTSTACK);
+                case NodeType::WRAP_A: return Cat(Cat((CScript() << OP_TOALTSTACK), std::move(subs[0])), (CScript() << OP_FROMALTSTACK));
                 case NodeType::WRAP_S: return (CScript() << OP_SWAP) + std::move(subs[0]);
                 case NodeType::WRAP_C: return std::move(subs[0]) + CScript() << (verify ? OP_CHECKSIGVERIFY : OP_CHECKSIG);
                 case NodeType::WRAP_D: return (CScript() << OP_DUP << OP_IF) + std::move(subs[0]) + (CScript() << OP_ENDIF);
```

I'll look into adding a `Cat()` which takes a variadic arguments, instead of nesting them and creating more vectors than needed.

taveechaimekwan replied:
220536

jb55 commented:
nice. Concept ACK.

dunxen commented:
Concept ACK

darosior commented:
> Concept ACK. I wonder if it makes sense to start out with a small subset of miniscript, so reviewers can focus on the implementation rather than on completeness.

I have tried to split this work into as many PR as reasonable, to reduce the size of this first chunk as much as possible. Note that 21% (you can't make this up) of the diff here is tests, and even more so in the following PRs. For smaller chunks to review, commits here should be focused, atomic and independently reviewable.

Further, i'm currently simplifying a few things following Pieter's comments above.
I'm happy to help review as much as i can, and if not for the first reviewers, comments like "it took time for me to grasp this was doing X, [adding a comment here / rewriting it this way] would be helpful to following reviewers" are very welcome!

Sjors commented:
It's mostly 3a2c00d9bc3c7a34fb8727d76e9c3a691395b67c that introduces a rather large body of stuff. Conceptually `c:pk_k(key)` is the easiest thing to understand, the equivalent of a `pk()` descriptor. So one commit could introduce `NodeType` with just `PK_K` and `WRAP_C`, `Type` with just B and K implemented. That should make that commit about 90% smaller, while still introducing the main moving parts of a parser, constraint checking, etc.

achow101 commented on src/script/miniscript.h:
```diff
@@ -0,0 +1,998 @@
+            } else if (Const("1", in)) {
+                constructed.push_back(MakeNodeRef<Key>(NodeType::JUST_1));
+            } else if (Const("pk(", in)) {
+                Key key;
+                int key_size = FindNextChar(in, ')');
+                if (key_size < 1) return {};
+                if (!ctx.FromString(in.begin(), in.begin() + key_size, key)) return {};
+                constructed.push_back(MakeNodeRef<Key>(NodeType::WRAP_C, Vector(MakeNodeRef<Key>(NodeType::PK_K, Vector(std::move(key))))));
+                in = in.subspan(key_size + 1);
```
In 3a2c00d9bc3c7a34fb8727d76e9c3a691395b67c "Miniscript: type system, script creation, text notation, tests"

Can the key parsing stuff be refactored into its own function instead of copy and pasted for the next 3 expressions?

achow101 commented on src/script/miniscript.h:
```diff
@@ -0,0 +1,998 @@
+                constructed.push_back(MakeNodeRef<Key>(NodeType::PK_H, Vector(std::move(key))));
+                in = in.subspan(key_size + 1);
+            } else if (Const("sha256(", in)) {
+                int hash_size = FindNextChar(in, ')');
+                if (hash_size < 1) return {};
+                std::string val = std::string(in.begin(), in.begin() + hash_size);
+                if (!IsHex(val)) return {};
+                auto hash = ParseHex(val);
+                if (hash.size() != 32) return {};
+                constructed.push_back(MakeNodeRef<Key>(NodeType::SHA256, std::move(hash)));
+                in = in.subspan(hash_size + 1);
```
In 3a2c00d9bc3c7a34fb8727d76e9c3a691395b67c "Miniscript: type system, script creation, text notation, tests"

Can the hex data parsing stuff be refactored into a function instead of copy and pasted for the next 3 expressions?

achow101 commented on src/script/miniscript.cpp:
```diff
@@ -0,0 +1,295 @@
+
+size_t ComputeScriptLen(NodeType nodetype, Type sub0typ, size_t subsize, uint32_t k, size_t n_subs, size_t n_keys) {
+    switch (nodetype) {
+        case NodeType::PK_K: return subsize + 34;
+        case NodeType::PK_H: return subsize + 3 + 21;
+        case NodeType::OLDER: return subsize + 1 + (CScript() << k).size();
+        case NodeType::AFTER: return subsize + 1 + (CScript() << k).size();
+        case NodeType::HASH256: return subsize + 4 + 2 + 33;
+        case NodeType::HASH160: return subsize + 4 + 2 + 21;
+        case NodeType::SHA256: return subsize + 4 + 2 + 33;
+        case NodeType::RIPEMD160: return subsize + 4 + 2 + 21;
+        case NodeType::WRAP_A: return subsize + 2;
+        case NodeType::WRAP_S: return subsize + 1;
+        case NodeType::WRAP_C: return subsize + 1;
+        case NodeType::WRAP_D: return subsize + 3;
+        case NodeType::WRAP_V: return subsize + (sub0typ << "x"_mst);
+        case NodeType::WRAP_J: return subsize + 4;
+        case NodeType::WRAP_N: return subsize + 1;
+        case NodeType::JUST_1: return 1;
+        case NodeType::JUST_0: return 1;
+        case NodeType::AND_V: return subsize;
+        case NodeType::AND_B: return subsize + 1;
+        case NodeType::OR_B: return subsize + 1;
+        case NodeType::OR_D: return subsize + 3;
+        case NodeType::OR_C: return subsize + 2;
+        case NodeType::OR_I: return subsize + 3;
+        case NodeType::ANDOR: return subsize + 3;
+        case NodeType::THRESH: return subsize + n_subs + 1;
+        case NodeType::MULTI: return subsize + 3 + (n_keys > 16) + (k > 16) + 34 * n_keys;
```
In 3a2c00d9bc3c7a34fb8727d76e9c3a691395b67c "Miniscript: type system, script creation, text notation, tests"

Several of these sizes are the same, so this could be deduplicated to be a bit easier to read.

darosior replied:
Also removing the `subsize` where it was not necessary (it must always be 0 for "leaf" fragments). I think it's much less confusing now.

achow101 commented on src/test/miniscript_tests.cpp:
```diff
@@ -0,0 +1,407 @@
+
+/** A class encapsulating conversion routing for CPubKey. */
+struct KeyConverter {
+    typedef CPubKey Key;
```
In 3a2c00d9bc3c7a34fb8727d76e9c3a691395b67c "Miniscript: type system, script creation, text notation, tests"

This appears to be unused.

darosior replied:
It is? See the `CONVERTER` global.

achow101 replied:
Oh, I grepped for `Key` and did not see it, but removing this does indeed cause a compilation failure.

achow101 commented on src/script/miniscript.h:
```diff
@@ -0,0 +1,998 @@
+    constexpr Type operator&(Type x) const { return Type(m_flags & x.m_flags); }
+
+    //! Check whether the left hand's properties are superset of the right's (= left is a subtype of right).
+    constexpr bool operator<<(Type x) const { return (x.m_flags & ~m_flags) == 0; }
```
In 3a2c00d9bc3c7a34fb8727d76e9c3a691395b67c "Miniscript: type system, script creation, text notation, tests"

`<<` confused me for a while as I am used to it meaning bit shift when used in a flags like context, rather than a containment operator.

willcl-ark commented on src/script/miniscript.h:
```diff
@@ -0,0 +1,1561 @@
+ * - "g" Whether the branch contains a relative time timelock
+ * - "h" Whether the branch contains a relative height timelock
+ * - "i" Whether the branch contains a absolute time timelock
+ * - "j" Whether the branch contains a absolute time heightlock
```
If you touch this perhaps worth changing to "absolute height timelock"

flack replied:
mini-nit: It should be "a**n** absolute [...]" in both lines

darosior commented on src/script/miniscript.h:
```diff
@@ -0,0 +1,998 @@
+        // and the CScripts of its child nodes, the CScript of the node.
+        auto upfn = [&ctx](bool verify, const Node& node, Span<CScript> subs) -> CScript {
+            switch (node.nodetype) {
+                case NodeType::PK_K: return CScript() << ctx.ToPKBytes(node.keys[0]);
+                case NodeType::PK_H: return CScript() << OP_DUP << OP_HASH160 << ctx.ToPKHBytes(node.keys[0]) << OP_EQUALVERIFY;
+                case NodeType::OLDER: return CScript() << node.k << OP_CHECKSEQUENCEVERIFY;
+                case NodeType::AFTER: return CScript() << node.k << OP_CHECKLOCKTIMEVERIFY;
+                case NodeType::SHA256: return CScript() << OP_SIZE << 32 << OP_EQUALVERIFY << OP_SHA256 << node.data << (verify ? OP_EQUALVERIFY : OP_EQUAL);
+                case NodeType::RIPEMD160: return CScript() << OP_SIZE << 32 << OP_EQUALVERIFY << OP_RIPEMD160 << node.data << (verify ? OP_EQUALVERIFY : OP_EQUAL);
+                case NodeType::HASH256: return CScript() << OP_SIZE << 32 << OP_EQUALVERIFY << OP_HASH256 << node.data << (verify ? OP_EQUALVERIFY : OP_EQUAL);
+                case NodeType::HASH160: return CScript() << OP_SIZE << 32 << OP_EQUALVERIFY << OP_HASH160 << node.data << (verify ? OP_EQUALVERIFY : OP_EQUAL);
+                case NodeType::WRAP_A: return (CScript() << OP_TOALTSTACK) + std::move(subs[0]) + (CScript() << OP_FROMALTSTACK);
+                case NodeType::WRAP_S: return (CScript() << OP_SWAP) + std::move(subs[0]);
+                case NodeType::WRAP_C: return std::move(subs[0]) + CScript() << (verify ? OP_CHECKSIGVERIFY : OP_CHECKSIG);
+                case NodeType::WRAP_D: return (CScript() << OP_DUP << OP_IF) + std::move(subs[0]) + (CScript() << OP_ENDIF);
+                case NodeType::WRAP_V: return std::move(subs[0]) + (node.subs[0]->GetType() << "x"_mst ? (CScript() << OP_VERIFY) : CScript());
+                case NodeType::WRAP_J: return (CScript() << OP_SIZE << OP_0NOTEQUAL << OP_IF) + std::move(subs[0]) + (CScript() << OP_ENDIF);
+                case NodeType::WRAP_N: return std::move(subs[0]) + CScript() << OP_0NOTEQUAL;
+                case NodeType::JUST_1: return CScript() << OP_1;
+                case NodeType::JUST_0: return CScript() << OP_0;
+                case NodeType::AND_V: return std::move(subs[0]) + std::move(subs[1]);
+                case NodeType::AND_B: return std::move(subs[0]) + std::move(subs[1]) + (CScript() << OP_BOOLAND);
+                case NodeType::OR_B: return std::move(subs[0]) + std::move(subs[1]) + (CScript() << OP_BOOLOR);
+                case NodeType::OR_D: return std::move(subs[0]) + (CScript() << OP_IFDUP << OP_NOTIF) + std::move(subs[1]) + (CScript() << OP_ENDIF);
+                case NodeType::OR_C: return std::move(subs[0]) + (CScript() << OP_NOTIF) + std::move(subs[1]) + (CScript() << OP_ENDIF);
+                case NodeType::OR_I: return (CScript() << OP_IF) + std::move(subs[0]) + (CScript() << OP_ELSE) + std::move(subs[1]) + (CScript() << OP_ENDIF);
+                case NodeType::ANDOR: return std::move(subs[0]) + (CScript() << OP_NOTIF) + std::move(subs[2]) + (CScript() << OP_ELSE) + std::move(subs[1]) + (CScript() << OP_ENDIF);
```
Yes, but it is going to go along with https://github.com/bitcoin/bitcoin/pull/24147#discussion_r795211395 in favour of a `BuildScript(OP, sub, sub, OP, OP, sub, ...)`.

darosior commented on src/script/miniscript.h:
```diff
@@ -0,0 +1,998 @@
+                    break;
+                default: break;
+            }
+            switch (node.nodetype) {
```
Note the duplicated case between the two switches. Some nodes' text representation can be abbreviated using a wrapper, and you want to try that first.
Besides that, i guess it's a matter of style between trying all wrappers first and merging them between the respective node types that are currently duplicated.

Sjors commented:
I peeled back fragments like an onion, giving each deleted fragment it's own commit. The order is somewhat random, but I started with the ones that are not needed for descriptors. I then reverted all these commits, and squashed the original ones. The end result is a series of commits that builds all of Miniscript piecemeal: https://github.com/Sjors/bitcoin/commits/2022/02/miniscript

If you want to use it, you should be able to rebase the rest of this PR on top of it.

The types and properties are now introduced in the commit for the first fragment that needs them. Because I peeled back in somewhat random order, this may not be the most optimal way to introduce them. But at least they're not all introduced at once. That said, the bulk of properties is still introduced in the first commit, because c:pk_h() uses the kitchensink.

Only commits d9d8bb044ffb62d5b7169d2e17859ce6754f5c60, 27cf992768c21bd0cd18e06fae242bc356be4c37 and 9bd615d2c99a77d29416891b7e17413dcf21ed28 are needed for our current descriptor capability.

The other commits implement the rest of miniscript could be moved to another PR. They can be reviewed in parallel with your followup PR's that integrate miniscript with descriptors.

I put `older()` and `after()` immediately after that in 6f87083fd6b01f3f7a5228a05702c0177dfe1c75, since time/height locks are probably the most useful wallet feature to introduce next (and the wallet RPC already has some timelock related functionality). That said,  it's probably fine to introduce all the new miniscript functionality in one go.

While peeling back fragments I was pretty reckless with the tests: I just deleted any test that contained the fragment. This leads to low coverage for the initial commits, which can probably be improved with some new tests.

darosior commented on src/script/miniscript.h:
```diff
@@ -0,0 +1,998 @@
+NodeRef<Key> MakeNodeRef(Args&&... args) { return std::make_shared<const Node<Key>>(std::forward<Args>(args)...); }
+
+//! The different node types in miniscript.
+enum class NodeType {
```
I'm renaming `NodeType` to `Fragment`.

achow101 commented on src/script/miniscript.h:
```diff
@@ -555,10 +600,110 @@ struct Node {
 
+    internal::Ops CalcOps() const {
+        switch (nodetype) {
+            case Fragment::PK_K: return {0, 0, 0};
+            case Fragment::PK_H: return {3, 0, 0};
+            case Fragment::OLDER: return {1, 0, {}};
+            case Fragment::AFTER: return {1, 0, {}};
+            case Fragment::SHA256: return {4, 0, {}};
+            case Fragment::RIPEMD160: return {4, 0, {}};
+            case Fragment::HASH256: return {4, 0, {}};
+            case Fragment::HASH160: return {4, 0, {}};
+            case Fragment::AND_V: return {subs[0]->ops.stat + subs[1]->ops.stat, subs[0]->ops.sat + subs[1]->ops.sat, {}};
+            case Fragment::AND_B: return {1 + subs[0]->ops.stat + subs[1]->ops.stat, subs[0]->ops.sat + subs[1]->ops.sat, subs[0]->ops.dsat + subs[1]->ops.dsat};
+            case Fragment::OR_B: return {1 + subs[0]->ops.stat + subs[1]->ops.stat, Choose(subs[0]->ops.sat + subs[1]->ops.dsat, subs[1]->ops.sat + subs[0]->ops.dsat), subs[0]->ops.dsat + subs[1]->ops.dsat};
+            case Fragment::OR_D: return {3 + subs[0]->ops.stat + subs[1]->ops.stat, Choose(subs[0]->ops.sat, subs[1]->ops.sat + subs[0]->ops.dsat), subs[0]->ops.dsat + subs[1]->ops.dsat};
+            case Fragment::OR_C: return {2 + subs[0]->ops.stat + subs[1]->ops.stat, Choose(subs[0]->ops.sat, subs[1]->ops.sat + subs[0]->ops.dsat), {}};
+            case Fragment::OR_I: return {3 + subs[0]->ops.stat + subs[1]->ops.stat, Choose(subs[0]->ops.sat, subs[1]->ops.sat), Choose(subs[0]->ops.dsat, subs[1]->ops.dsat)};
+            case Fragment::ANDOR: return {3 + subs[0]->ops.stat + subs[1]->ops.stat + subs[2]->ops.stat, Choose(subs[1]->ops.sat + subs[0]->ops.sat, subs[0]->ops.dsat + subs[2]->ops.sat), subs[0]->ops.dsat + subs[2]->ops.dsat};
+            case Fragment::MULTI: return {1, (uint32_t)keys.size(), (uint32_t)keys.size()};
+            case Fragment::WRAP_A: return {2 + subs[0]->ops.stat, subs[0]->ops.sat, subs[0]->ops.dsat};
+            case Fragment::WRAP_S: return {1 + subs[0]->ops.stat, subs[0]->ops.sat, subs[0]->ops.dsat};
+            case Fragment::WRAP_C: return {1 + subs[0]->ops.stat, subs[0]->ops.sat, subs[0]->ops.dsat};
+            case Fragment::WRAP_D: return {3 + subs[0]->ops.stat, subs[0]->ops.sat, 0};
+            case Fragment::WRAP_V: return {subs[0]->ops.stat + (subs[0]->GetType() << "x"_mst), subs[0]->ops.sat, {}};
+            case Fragment::WRAP_J: return {4 + subs[0]->ops.stat, subs[0]->ops.sat, 0};
+            case Fragment::WRAP_N: return {1 + subs[0]->ops.stat, subs[0]->ops.sat, subs[0]->ops.dsat};
+            case Fragment::JUST_1: return {0, 0, {}};
+            case Fragment::JUST_0: return {0, {}, 0};
```
In 85cf12e20201eee4ed23902e1da151f63663cfab "Miniscript: ops limit and stack size computation"

A lot of these are the same, so this could be de-duplicated in the same way that the size calculation was.

achow101 commented on src/script/miniscript.h:
```diff
@@ -555,10 +600,110 @@ struct Node {
+    }
+
+    internal::StackSize CalcStackSize() const {
+        switch (nodetype) {
+            case Fragment::PK_K: return {1, 1};
+            case Fragment::PK_H: return {2, 2};
+            case Fragment::OLDER: return {0, {}};
+            case Fragment::AFTER: return {0, {}};
+            case Fragment::SHA256: return {1, {}};
+            case Fragment::RIPEMD160: return {1, {}};
+            case Fragment::HASH256: return {1, {}};
+            case Fragment::HASH160: return {1, {}};
+            case Fragment::ANDOR: return {Choose(subs[0]->ss.sat + subs[1]->ss.sat, subs[0]->ss.dsat + subs[2]->ss.sat), subs[0]->ss.dsat + subs[2]->ss.dsat};
+            case Fragment::AND_V: return {subs[0]->ss.sat + subs[1]->ss.sat, {}};
+            case Fragment::AND_B: return {subs[0]->ss.sat + subs[1]->ss.sat, subs[0]->ss.dsat + subs[1]->ss.dsat};
+            case Fragment::OR_B: return {Choose(subs[0]->ss.dsat + subs[1]->ss.sat, subs[0]->ss.sat + subs[1]->ss.dsat), subs[0]->ss.dsat + subs[1]->ss.dsat};
+            case Fragment::OR_C: return {Choose(subs[0]->ss.sat, subs[0]->ss.dsat + subs[1]->ss.sat), {}};
+            case Fragment::OR_D: return {Choose(subs[0]->ss.sat, subs[0]->ss.dsat + subs[1]->ss.sat), subs[0]->ss.dsat + subs[1]->ss.dsat};
+            case Fragment::OR_I: return {Choose(subs[0]->ss.sat + 1, subs[1]->ss.sat + 1), Choose(subs[0]->ss.dsat + 1, subs[1]->ss.dsat + 1)};
+            case Fragment::MULTI: return {(uint32_t)keys.size() + 1, (uint32_t)keys.size() + 1};
+            case Fragment::WRAP_A: return subs[0]->ss;
+            case Fragment::WRAP_S: return subs[0]->ss;
+            case Fragment::WRAP_C: return subs[0]->ss;
+            case Fragment::WRAP_D: return {1 + subs[0]->ss.sat, 1};
+            case Fragment::WRAP_V: return {subs[0]->ss.sat, {}};
+            case Fragment::WRAP_J: return {subs[0]->ss.sat, 1};
+            case Fragment::WRAP_N: return subs[0]->ss;
+            case Fragment::JUST_1: return {0, {}};
+            case Fragment::JUST_0: return {{}, 0};
```
In 85cf12e20201eee4ed23902e1da151f63663cfab "Miniscript: ops limit and stack size computation"

A lot of these are the same, so this can be de-duplicated in the same way that the size calculation was.

achow101 commented on src/script/miniscript.h:
```diff
@@ -555,10 +600,110 @@ struct Node {
+        switch (nodetype) {
+            case Fragment::PK_K: return {0, 0, 0};
+            case Fragment::PK_H: return {3, 0, 0};
+            case Fragment::OLDER: return {1, 0, {}};
```
In 85cf12e20201eee4ed23902e1da151f63663cfab "Miniscript: ops limit and stack size computation"

What is the distinction between `{}` and `0`? How com some fragments use 0, and others use `{}`?

It seems like `{}` is being used for fragments that have no dis/satisfaction. However it is used for the hashing fragments and those have a dissatisfaction.

sipa replied:
While hashes do have dissatisfactions, they are always malleable. The ops counting algorithm only works for non-malleable signing, which will never use these.

achow101 commented on src/script/miniscript.h:
```diff
@@ -555,10 +600,110 @@ struct Node {
+    bool CheckOpsLimit() const { return GetOps() <= MAX_OPS_PER_SCRIPT; }
+
+    //! Return the maximum number of stack elements needed to satisfy this script non-malleably.
+    uint32_t GetStackSize() const { return ss.sat.value + 1; }
```
In 85cf12e20201eee4ed23902e1da151f63663cfab "Miniscript: ops limit and stack size computation"

It was not immediately obvious to me that this includes the script itself as a stack item, perhaps mention that in the docstring?

darosior commented on src/script/miniscript.h:
```diff
@@ -231,6 +231,47 @@ size_t ComputeScriptLen(Fragment nodetype, Type sub0typ, size_t subsize, uint32_
+
+struct Ops {
+    //! Non-push opcodes.
+    uint32_t stat;
```
Naming a field of `Ops` "`ops`" is a bit weird, went with `count` instead.

darosior commented on src/script/miniscript.h:
```diff
@@ -1007,13 +1008,429 @@ inline NodeRef<Key> Parse(Span<const char> in, const Ctx& ctx)
+                constructed.push_back(MakeNodeRef<Key>(Fragment::AFTER, k));
+            }
+            // Hashes
+            else if (last - in >= 7 && in[0].first == OP_EQUAL && in[1].second.size() == 32 && in[2].first == OP_SHA256 && in[3].first == OP_VERIFY && in[4].first == OP_EQUAL && ParseScriptNumber(in[5], k) && k == 32 && in[6].first == OP_SIZE) {
+                constructed.push_back(MakeNodeRef<Key>(Fragment::SHA256, in[1].second));
+                in += 7;
+            } else if (last - in >= 7 && in[0].first == OP_EQUAL && in[1].second.size() == 20 && in[2].first == OP_RIPEMD160 && in[3].first == OP_VERIFY && in[4].first == OP_EQUAL && ParseScriptNumber(in[5], k) && k == 32 && in[6].first == OP_SIZE) {
+                constructed.push_back(MakeNodeRef<Key>(Fragment::RIPEMD160, in[1].second));
+                in += 7;
+            } else if (last - in >= 7 && in[0].first == OP_EQUAL && in[1].second.size() == 32 && in[2].first == OP_HASH256 && in[3].first == OP_VERIFY && in[4].first == OP_EQUAL && ParseScriptNumber(in[5], k) && k == 32 && in[6].first == OP_SIZE) {
+                constructed.push_back(MakeNodeRef<Key>(Fragment::HASH256, in[1].second));
+                in += 7;
+            } else if (last - in >= 7 && in[0].first == OP_EQUAL && in[1].second.size() == 20 && in[2].first == OP_HASH160 && in[3].first == OP_VERIFY && in[4].first == OP_EQUAL && ParseScriptNumber(in[5], k) && k == 32 && in[6].first == OP_SIZE) {
```
If i do that i need to get rid of all the `else if`s and replace them with `if { ...; break;}`. I can do that but it's more a matter of taste than a net benefit.

darosior commented:
Addressed Andrew's comments, i also reworked the fuzz target. As a consequence i added a couple missing checks in the function decoding from Script. We are currently working on different approaches to write one or more fuzz targets generating random fragments in place of the random unit tests, which was removed in the previous push. They could be included here or in #24149 since it's tightly related to testing the satisfaction logic.

Regarding Sjors' point about splitting commits further i'd like to have other reviewers' opinion. I'm personally still not sure that it makes sense but heh i'm not the one reviewing the PR :). So if others think that would be useful, i could break down the first commit into tiny ones..

achow101 commented:
ACK 4b74cc97d3543efe6c2f126da5e807da72cc782c

achow101 commented on src/script/miniscript.h:
```diff
@@ -555,10 +600,148 @@ struct Node {
+            case Fragment::WRAP_J: return {4 + subs[0]->ops.count, subs[0]->ops.sat, 0};
+            case Fragment::WRAP_V: return {subs[0]->ops.count + (subs[0]->GetType() << "x"_mst), subs[0]->ops.sat, {}};
+            case Fragment::THRESH: {
+                uint32_t count = 0;
+                auto sats = Vector(internal::MaxInt<uint32_t>(0));
+                for (const auto& sub : subs) {
+                    count += sub->ops.count + 1;
+                    auto next_sats = Vector(sats[0] + sub->ops.dsat);
+                    for (size_t j = 1; j < sats.size(); ++j) next_sats.push_back(Choose(sats[j] + sub->ops.dsat, sats[j - 1] + sub->ops.sat));
+                    next_sats.push_back(sats[sats.size() - 1] + sub->ops.sat);
+                    sats = std::move(next_sats);
+                }
+                assert(k <= sats.size());
+                return {count, sats[k], sats[0]};
```
in f38df6abfb40120899c7324b7e690d1b585bc26b "Miniscript: ops limit and stack size computation"

I'm having a bit of a hard time figuring out how this arrives at the right answer. Could you provide a brief explanation of how this loop works to count the opcodes (and stack size, and the input stack, in other places)?

darosior replied:
The algorithm used here is the same as the one used for the satisfaction of a thresh fragment in #24149. The intent is to compute the maximum satisfaction and dissatisfaction cost (in various resources) for a given fragment.

For a `thresh` with `n` sub-fragments and a threshold of `k` the algorithm will end up with a vector of `n` elements, acting as a mapping from `k` -> cost. With the special case of `k == 0` considered the dissatisfaction of the `thresh` fragment ([the only non-malleable dissatisfaction of this fragment](https://bitcoin.sipa.be/miniscript/)). Note a `thresh` with a `k` of 0 is invalid.

The algorithm works by adding for each sub the cost of its dissatisfaction to the first entry of the vector, and the cost of its satisfaction to the last entry.
The entries from `1` to `n - 1` are then updated with the bigger of:
- the existing cost (at `k`) plus the dissatisfaction cost of the current sub, that is we have enough subs satisfied and we dissatisfy the current sub
- the previous cost (`k - 1`) plus the satisfaction cost of the current sub, that is we are missing one sub to get to the threshold and we satisfy the current sub

darosior commented:
Pushed some cleanups by @sipa (https://github.com/sipa/miniscript/pull/101).

sanket1729 commented on src/script/miniscript.h:
```diff
@@ -556,10 +601,148 @@ struct Node {
+            case Fragment::OR_C: return {Choose(subs[0]->ss.sat, subs[0]->ss.dsat + subs[1]->ss.sat), {}};
+            case Fragment::OR_D: return {Choose(subs[0]->ss.sat, subs[0]->ss.dsat + subs[1]->ss.sat), subs[0]->ss.dsat + subs[1]->ss.dsat};
+            case Fragment::OR_I: return {Choose(subs[0]->ss.sat + 1, subs[1]->ss.sat + 1), Choose(subs[0]->ss.dsat + 1, subs[1]->ss.dsat + 1)};
+            case Fragment::MULTI: return {(uint32_t)keys.size() + 1, (uint32_t)keys.size() + 1};
```
In 65c79db36818acc193e21c4b5bb6bf7dbf8ff764:

This should be a number of signatures instead of `keys`.

darosior replied:
Done, and added a unit test reproducing this.

sanket1729 reviewed (commented):
Did a partial review. Will review the remaining PR shortly.

fanquake commented:
@stepansnigirev you might be interested in reviewing this.

sipa commented:
We'll address all comments that have been left here soon, but just FYI, there are a bunch of improvements still ongoing in the miniscript repo itself (https://github.com/sipa/miniscript/, most of which will eventually be moved here after this PR and the followups).

darosior commented on src/script/miniscript.h:
```diff
@@ -0,0 +1,998 @@
+ *   - Cannot be dissatisfied.
+ *   - This is obtained by adding an OP_VERIFY to a B, modifying the last opcode
+ *     of a B to its -VERIFY version (only for OP_CHECKSIG, OP_CHECKSIGVERIFY
+ *     and OP_EQUAL), or using IFs where both branches are also Vs.
```
I've updated the comment to be more general but not exhaustive:
```diff
diff --git a/src/script/miniscript.h b/src/script/miniscript.h
index 88270e54d7..070dcc9d14 100644
--- a/src/script/miniscript.h
+++ b/src/script/miniscript.h
@@ -42,9 +42,9 @@ namespace miniscript {
  *   - Takes its inputs from the top of the stack.
  *   - When satisfactied, pushes nothing.
  *   - Cannot be dissatisfied.
- *   - This is obtained by adding an OP_VERIFY to a B, modifying the last opcode
+ *   - This can be obtained by adding an OP_VERIFY to a B, modifying the last opcode
  *     of a B to its -VERIFY version (only for OP_CHECKSIG, OP_CHECKSIGVERIFY
- *     and OP_EQUAL), or using IFs where both branches are also Vs.
+ *     and OP_EQUAL), or by combining a V fragment under some conditions.
  *   - For example vc:pk_k(key) = <key> OP_CHECKSIGVERIFY
  * - "K" Key:
  *   - Takes its inputs from the top of the stack.
```

darosior committed:
script: expose getter for CScriptNum, add a BuildScript helper

Some prep work for Miniscript. BuildScript is an efficient way to build
Scripts in a generic manner (by concatenating OPs, data, and other
Scripts).

Co-Authored-By: Pieter Wuille <pieter@wuille.net>

darosior commented:
Rebased. This is now ready for review.

There are still refactorings to the fuzz targets that have not landed in the Miniscript repo yet, i'll either add them here eventually or we'll open them as followup PRs. See https://github.com/sipa/miniscript/issues/98 for details.

Changes since last push:
- Fixed the `CHECKMULTISIG` stack size calculation bug (thanks Sanket!)
- `Choose` was renamed to `operator|`
- The modifications to `script.h` were split from the type system commit to make it only introduce `miniscript.h`
- Some leftovers (from previous iterations of concatenating Scripts) were removed from `vector.h`

jb55 commented:
ACK 2da94a4c6f55f7a3621f4a6f70902c52f735c868

I've been running the miniscript_decode fuzzer for awhile and it looks good. looking forward to testing the smarter fuzzers after this is merged.

laanwj commented on src/script/miniscript.cpp:
```diff
@@ -0,0 +1,348 @@
+    int num_types = (e << "K"_mst) + (e << "V"_mst) + (e << "B"_mst) + (e << "W"_mst);
+    if (num_types == 0) return ""_mst; // No valid type, don't care about the rest
+    assert(num_types == 1); // K, V, B, W all conflict with each other
+    bool ok = // Work around a GCC 4.8 bug that breaks user-defined literals in macro calls.
```
FYI: we require gcc 8.1 according to `doc/dependencies.md`. There's no need to work around gcc 4.8 bugs.

sipa replied:
See https://github.com/sipa/miniscript/pull/112

darosior commented:
Reviewers, before the integration into output descriptors, i've made a follow-up PR with the final updates and cleanups we made in the Miniscript repo.
https://github.com/bitcoin/bitcoin/pull/24860

sanket1729 commented on src/script/miniscript.h:
```diff
@@ -556,10 +601,148 @@ struct Node {
+            case Fragment::SHA256:
+            case Fragment::RIPEMD160:
+            case Fragment::HASH256:
+            case Fragment::HASH160: return {1, {}};
```
This should be {1, 0}. All hash types can be dis-satisfied with a random 32-byte pre-image.

darosior replied:
But it's malleable, so we don't want to account for it.

sanket1729 replied:
Why be different while calculating stack size and opcodes? In calculating executed op code count, we count the correct opcodes when this expression is dissatisfied. But we don't do so while considering stack size?

sipa replied:
The hashing fragments return `{4, 0, {}}` in CalcOps, and `{1, {}}` in CalcStackSize? In both cases, no value (`{}`) is returned for the dissatisfaction case, because the non-malleable signing algorithm will never use the dissatisfaction path.

sanket1729 replied:
Sorry, error on my part. You are correct

sanket1729 replied:
@sipa @darosior Revisiting this: 
I am considering should malleability rules even come into play over here? Any witness malleability cannot change the max opcode count. Though, it can potentially change the witness stack element count. 

An example test-case in rust-miniscript broke of this,

> `c:and_v(or_c(sha256(H),v:multi(1,A)),pk_k(B))`

Outputs `maxops=8` because it does not count the malleable solution we dissatisfy `sha256`. Just trying to stay consistent with the definition of OpsCount with rust-miniscript.

If we are ignoring malleable solutions in `Ops/StackElems`, how about we change `GetOps/GetStackSize` to output Error/Option/-1 when it operates on the malleable script?

sipa replied:
@sanket1729 That's probably reasonable, but I'm not sure it's needed.

Really what these functions (as currently written) answer is: "what is the maximum stacksize/opcode needed for non-malleable satisfactions". If you have a script with branches that cannot be satisfied non-malleable, you probably don't care about the numbers of just the non-malleable ones, but you also probably don't care about these functions in general. Their purpose is really only contributing to determining IsSane(TopLevel), and in case of not IsNonMalleable(), IsSane() will already be false.

sanket1729 replied:
@sipa Agreed. It might not be needed. I was translating the test vector from this PR to rust-miniscript when I noted the difference between the implementations. 

> c:and_v(or_c(sha256(H),v:multi(1,A)),pk_k(B)) with MaxOps() = 8
but rust-miniscript outputs 9.

I was also using them for cross-testing outputs with rust-miniscript to check why the value mismatches here. I agree that from an end-user perspective this is not really needed. 

It might not be worth it to change the tests for future miniscript implementations(if any).

sanket1729 commented on src/script/miniscript.h:
```diff
@@ -556,10 +601,148 @@ struct Node {
+            }
+        }
+        assert(false);
+        return {{}, {}};
```
What's the purpose of return after an assert? Is this normal in c++? Same comment for the above function

darosior replied:
Yeah it's not necessary, Marco mentioned it to me elsewhere. I can remove it in #24860

sanket1729 reviewed (commented):
Finally completed the review. Sorry for the delay. Did a line-by-line comparison of all miniscript code with rust-miniscript. Did not review tests/fuzzer. 

Found a couple of issues that can be addressed in followups. ACK for the rest of it

stickies-v commented on src/script/miniscript.h:
```diff
@@ -0,0 +1,1652 @@
+    //! Compute the type with the intersection of properties.
+    constexpr Type operator&(Type x) const { return Type(m_flags & x.m_flags); }
+
+    //! Check whether the left hand's properties are superset of the right's (= left is a subtype of right).
```
Isn't it the other way around, where if `true` then the right hand's properties are a superset of the left's?

stickies-v replied:
Sorry of course right after posting I see that the order of `m_flags` and `x.m_flags` is switched up for this operator compared to the other operators, which makes the docstring correct so please disregard my previous comment. The inconsistent order is maybe a bit confusing but no problem.

sipa replied:
No, the comment is correct I believe.

(x.m_flags & ~m_flags) is zero when every flag set in x.m_flags is also set in m_flags. In other words, where this's flags (=left hand) are a superset of x's flags.

darosior commented on src/script/miniscript.h:
```diff
@@ -556,10 +601,148 @@ struct Node {
     size_t ScriptSize() const { return scriptlen; }
 
+    //! Return the maximum number of ops needed to satisfy this script non-malleably.
+    uint32_t GetOps() const { return ops.count + ops.sat.value; }
```
True, but i'm not sure it's worth "fixing".
In this case returning `0` would be as wrong as any other value. We could probably make the return value `std::optional` but that sounds a bit overkill?

darosior commented on src/script/miniscript.cpp:
```diff
@@ -0,0 +1,295 @@
+        case Fragment::SHA256: return 4 + 2 + 33;
+        case Fragment::HASH160:
+        case Fragment::RIPEMD160: return 4 + 2 + 21;
+        case Fragment::MULTI: return 3 + (n_keys > 16) + (k > 16) + 34 * n_keys;
```
Done in https://github.com/bitcoin/bitcoin/pull/24860, thanks
